### DISCUSSION TITLE
How does I18N work for strings Theia brings in

### DISCUSSION THREAD

**Discussion by arunswarnam** [eclipse-theia/theia#11077] [Q&A] [answered]
Hi,
   Thank you for adding [I18N support](https://github.com/eclipse-theia/theia/pull/9538) and the [doc](https://theia-ide.org/docs/i18n). I was able to follow the doc and `nls-extract` the strings in my browser application. The extracted strings look like
   ```
{
    "vscode": {
        "bulkEdit.contribution": {
            "panel": "Refactor Preview",
            "apply": "Apply Refactoring",
            "Discard": "Discard Refactoring"
        },
        ....
        "textFileSaveErrorHandler": {
            "overwrite": "Overwrite"
        },
        "simpleFileDialog": {
            "remoteFileDialog.validateExisting": "Do you really want to overwrite \"{0}\"?"
        }
    },
    "theia": {
        "getting-started": {
            "newExtension": "Building a New Extension",
            "newPlugin": "Building a New Plugin"
        },
        "scm": {
            "amendHeadCommit": "HEAD Commit",
            "amendLastCommit": "Amend last commit",
            "amend": "Amend",
            "unamendCommit": "Unamend commit",
            "unamend": "Unamend",
            "changeRepository": "Change Repository..."
        },
    }
}
```

1. If I understand correctly string keys for vscode like `vscode/bulkEdit.contribution/panel` are localized from the [vscode language packs ](https://github.com/microsoft/vscode-loc/tree/main/i18n) I am curious from where the Theia strings keys like `theia/getting-started/newExtension` are localized?
2. I noticed Theia application fall-backs to custom localization provider for strings that are not localized by the vscode language packs. So, I was able to localize Theia strings mentioned above by adding a custom localization provider in our custom Theia extension but wasn't sure if this approach is recommended as the source of strings is a different npm package and the translation is going to be coming from a different npm package. Is there a recommendation on the best approach here?

**Comment by msujew** [comment, marked as answer]
Hi @arunswarnam,

1. Right, we source most of our localizations from vscode language packs. We use a utility method `nls.localizeByDefault` which allows us to skip entering vscode localization keys and only use the default value. The framework finds the correct key from the default value and then is able to translate it that way. This only works for stuff from vscode. The Theia translations are stored in [this directory](https://github.com/eclipse-theia/theia/tree/master/packages/core/i18n). The actual translations are done through the [DeepL API integration](https://github.com/eclipse-theia/theia/blob/master/dev-packages/localization-manager/src/deepl-api.ts).
2. There are no real best practices in that regard, as long as you use the recommended API outlined in the docs you are on the safe side (i.e. `LocalizationContribution` and its `register*` methods).

**Reply by arunswarnam** [reply]
Thank you @msujew
